[#50212] Convert file storage edit view to primer design

akabiru opened the pull request:
#### https://community.openproject.org/wp/50212

- [x] storage view
    - [x] Action buttons (edit, resync OAuth)
    - [x] Check with product on Cancel button on storage view ([work_packages/50499](https://community.openproject.org/projects/openproject/work_packages/50499))
    - 🚧  Fix Delete button (https://community.openproject.org/wp/50739)
    -  🚧 On Storage name, update `PageHeader` title (https://community.openproject.org/wp/50739)
- [x] Edit storage host form
- [x] Resync OpenProject OAuth credentials
    - ? TODO: Make `Done, continue` button context aware
- [x] Resync Nextcloud OAuth credentials
- [x] Managed project folders form

Primer Components
- [x] Extract `OpTurbo::FrameComponent` https://github.com/opf/openproject/pull/13932
- Create task for form input widths handling (https://community.openproject.org/wp/50740)
- [x] Create maintenance task to extract `OpPrimer::ClipboardCopyComponent` and primerise sizes (https://community.openproject.org/wp/50741)
- [x] Restrict storage switching in Edit page
- Adopt https://github.com/opf/openproject/pull/13952 once ready (https://community.openproject.org/wp/50742)


![demo-2](https://github.com/opf/openproject/assets/17295175/23ddc673-5d4d-45c8-8efa-63ce45c249b6)

akabiru committed:
chore[#50212]: Begin layout

akabiru committed:
chore[#50212]: Combine StorageButtonComponent with options

akabiru committed:
chore[#50212]: Rename to StorageComponent

akabiru committed:
Revert "chore[#50212]: Rename to StorageComponent"

This reverts commit 5d2cb41efed8c295442ac90d0cc474a7d855371d.

akabiru committed:
feat[#50212]: Retrieve status labels from configuration checks

akabiru committed:
fix[#50212]: Restore configuration checks component

akabiru committed:
fix[#50212]: Remove unused methods

akabiru committed:
chore[#50212]: edit_host_name_configuration blueprint

akabiru committed:
feat[#50212]: Define edit provider host form with validation

akabiru committed:
chore[#50212]: rename to `edit_host`

akabiru committed:
chore[#50212]: Define a `StorageResultErrorPresenter` that delegates model errors to the service

akabiru committed:
chore[#50212]: Namespace `Storage::Admin::Forms`

akabiru committed:
chore[#50212]: Remove redundant `storage_` prefix

akabiru committed:
chore[#50212]: Remove redundant unused helpers

akabiru committed:
feat[#50212]: Add `OpTurbo::FrameWrapperComponent`

akabiru committed:
chore[#50212]: Utilise FrameWrapperComponent in edit host view

aaron-contreras commented on spec/components/op_turbo/frame_wrapper_component_spec.rb:
```diff
@@ -0,0 +1,51 @@
+
+RSpec.describe OpTurbo::FrameWrapperComponent, type: :component do
+  describe '#turbo_frame_id' do
+    context 'with frame option' do
```
How about naming the option `:context` instead of `:frame`?

akabiru replied:
Yeah, that's a good shout 👍🏾

cbliard commented on app/components/op_turbo/frame_wrapper_component.rb:
```diff
@@ -0,0 +1,35 @@
+#++
+
+module OpTurbo
+  class FrameWrapperComponent < ApplicationComponent
```
I would remove the `Wrapper` part of the name.

akabiru replied:
Any reason why? I imagine we want to be consistent with `OpTurbo::StreamWrapperComponent`

aaron-contreras replied:
I think it's inline with `OpTurbo::StreamWrapperComponent`. They are serving the same underlying purpose so I'd either keep it with `Wrapper` or rename both. Thoughts?

cbliard replied:
True, I would remove `Wrapper` from `OpTurbo::StreamWrapperComponent` as well. 
When I read `OpTurbo::FrameWrapperComponent` and `OpTurbo::StreamWrapperComponent`, I'm thinking that they may be doing something special as they "wrap" something. But in fact no, they are just acting as expected.

I feel like "wrapper" is not adding any information here and the code reads equally well without it (and even better in my opinion). For instance when we are building a form, we are not using `form_wrapper_tag`, `label_wrapper_tag`, `input_wrapper_tag`, and `submit_wrapper_tag`. We are using `form_tag`, `label_tag`, `input_tag`, and `submit_tag` because they do what they announce: they render a tag of a particular kind. I think it is similar in this case: naming it `OpTurbo::FrameComponent` and `OpTurbo::StreamComponent` would make the name shorter while keeping the same intent.

cbliard replied:
That being said, it's ok for me if we keep the `Wrapper` part in the name.

akabiru replied:
Took up your recommendation @cbliard 👍🏾 extracted to a separate PR https://github.com/opf/openproject/pull/13932

akabiru committed:
chore[#50212]: WIP - OAuth Credentials clipboard copy

akabiru committed:
feat[#50212]: Add `OpenProject::Common::ClipboardCopyComponent`

akabiru committed:
chore[#50212]: Use clipboard copy component

akabiru committed:
chore[#50212]: Use primer text selector, set up basic spec

akabiru committed:
chore[#50212]: Add flash warning for credentials copy

akabiru committed:
chore[#50212]: Use attr options reader

akabiru committed:
fix[#50212]: Rename `OpTurbo::FrameWrapperComponent` `:frame` to `context`

https://github.com/opf/openproject/pull/13866/files#r1356978232

akabiru committed:
fix[#50212]: Adopt primer default to visually hide label in clipboard copy

akabiru committed:
Merge branch 'dev' into implementation/50212-convert-the-edit-page-to-primer-design

akabiru committed:
fix[Op#50212]: Minor improvements

akabiru committed:
feat[Op#50212]: Define OAuthClient Form flow

akabiru committed:
tests[Op#50212]: update basic view spec

akabiru committed:
chore[Op#50212]: Remove unused methods

akabiru committed:
chore[Op#50212]: combine common form inputs

akabiru committed:
chore[Op#50212]: remove unused forms

akabiru committed:
feat[Op#50212]: Complete OAuthClient form view

akabiru committed:
fix[Op#50212]: Align icon buttons at end in edit view

akabiru committed:
feat[Op#50212]: Add Automatically managed form view

akabiru committed:
fix[Op#50212]: Use correct labels

akabiru committed:
fix[Op#50212]: Add error presenter for oauth client

akabiru committed:
fix[Op#50212]: Decorate model with error presenter coditionally

akabiru committed:
fix[Op#50212]: Replace `turbo_confirm` with `confirm` which is more consistent

akabiru committed:
tests[Op#50212]: add tests for edit view

akabiru committed:
fix[Op#50212]: Let `NextcloudContract` fail with validation error when host is nil

akabiru committed:
tests[Op#50212]: Expand Storage "General information" spec

akabiru committed:
tests[Op#50212]: Expand Storage "OAuth applications" spec

akabiru committed:
fix[Op#50212]: Select field should be disabled in edit mode

akabiru committed:
tests[Op#50212]: Expand Storage "Nextcloud OAuth" spec

akabiru commented on app/components/open_project/common/clipboard_copy_component.rb:
```diff
@@ -0,0 +1,72 @@
+      options[:value_to_copy]
+    end
+  end
+end
```
- [x] Create maintenance task to extract `OpPrimer::ClipboardCopyComponent` and primerise sizes (https://community.openproject.org/wp/50741)

HDinger replied:
Hi @akabiru 

I am wondering why this whole component is necessary at all? Couldn't we use the trailing action of the text input to display the copy icon? And if it is needed, I'd strongly vote for a different name, as it is not a single copyToClipboardComponent but rather something like a textboxWithCopyToClipboardComponent.

HDinger replied:
Ok, I just double checked with how it is done with GitHub, and they are using an `input-group` which does not exist yet in Primer. So I'd vote for taking that name. I will update the mentioned ticket accordingly.

akabiru replied:
Discussed with Henriette- details added in https://community.openproject.org/wp/50741

akabiru commented on app/services/service_result_errors_presenter.rb:
```diff
@@ -0,0 +1,38 @@
+  end
+
+  delegate :errors, to: :service_result
+end
```
Adopt https://github.com/opf/openproject/pull/13952 once ready (https://community.openproject.org/wp/50742)

akabiru commented on modules/storages/app/controllers/storages/admin/storages_controller.rb:
```diff
@@ -179,7 +183,20 @@ def oauth_application(service_result)
       .permit('name', 'provider_type', 'host', 'oauth_client_id', 'oauth_client_secret')
   end
+
+  # TODO: Work out how to retrieve the storage provider resource name as it's based on the provider type
+  # PrimerForms implements Rails `form_with` which doesn't support overriding the form name as we would with
+  # `form_for`.
+  # See: https://github.com/opf/primer_view_components/blob/79fb58474771bd06946554f8325cd0b1bdd6dd31/app/helpers/primer/form_helper.rb#L7
+  #
+  def storage_provider_parameter_name
+    if OpenProject::FeatureDecisions.storage_primer_design_active?
+      :storages_nextcloud_storage
+    else
+      :storages_storage
+    end
+  end
```
@opf/storages started discussions of introducing a FormObject to represent model CRUD operations

https://community.openproject.org/wp/50743

akabiru committed:
tests[Op#50212]: Expand Storage "Automatically managed view" spec

Kharonus commented on modules/storages/app/components/storages/admin/storage_list_component.html.erb:
```diff
@@ -57,7 +57,7 @@
       component.with_description { I18n.t('storages.storage_list_blank_slate.description') }
 
-      component.with_primary_action(**Storages::Admin::NewStorageButtonComponent.new.button_options)  do |button|
+      component.with_primary_action(**Storages::Admin::StorageButtonComponent.new(action: :new).button_options)  do |button|
```
🟡 I have a small pain with this component. Let me try to explain:

The current state would imply, that we basically create a component for every(!) button we have out there. They are sometimes combinable, as you did here with the delete and create action. But overall, there will be one component per button.

Why can't we have just a button component, were I can give the icon and class (primary, secondary, danger) and a block that is to be executed on click?

Kharonus replied:
Though I of course agree, that we might want to reduce duplicate code, as the `+ Storage` button is used on different views, isn't it? But then again, I would not try to marry creation and deletion and instead just write two different components.

akabiru replied:
Yes, the motivation was to DRY up the `+ Storage` button- but I see your point that we probably don't want a component for every button. I'll split this up 👍🏾

Kharonus replied:
great, thanks a lot :D

Kharonus commented on modules/storages/config/locales/en.yml:
🟡 a lot of new additions. I wonder, if we can remove a good amount of the locale strings here after removing the feature flag introduced in this PR. Maybe worth mentioning in a follow up work package?

akabiru replied:
Yes! That is the plan, we could not reuse a lot of the existing translations unfortunately.

Noted in https://community.openproject.org/wp/50748 👍🏾

Kharonus commented on modules/storages/spec/contracts/storages/storages/nextcloud_contract_spec.rb:
```diff
@@ -55,14 +55,14 @@
 
+      before do
+        mock_server_capabilities_response(storage.host)
+        mock_server_config_check_response(storage.host)
```
🟢 Seeing those makes me think, if we shouldn't use the queries from within the contract, too. Then we have VCR testing covering the queries, and in this test we can simply stub the query response.

akabiru replied:
Yeah, that's a good point- I think that will be a good improvement that should simplify the contract tests 👍🏾

Kharonus commented on modules/storages/spec/features/admin_storages_spec.rb:
```diff
@@ -89,6 +89,156 @@
     end
   end
 
+  describe 'File storage edit view', with_flag: { storage_primer_design: true } do
```
🟢 nice work of rewriting the feature spec

Kharonus reviewed (approved):
In general I'm impressed by the work. good progress.

There is only a couple of issues I'd like to discuss or mention. But none of those are blockers for a merge from my pov.

akabiru committed:
Revert "chore[#50212]: Combine StorageButtonComponent with options"

This reverts commit 03dd1a03f9d6e3df177d0e897f4608112749ac37.

akabiru committed:
chore[Op#50212]: remove DeleteStorageButtonComponent, use Beta::Button directly

https://github.com/opf/openproject/pull/13866/commits/03dd1a03f9d6e3df177d0e897f4608112749ac37#r1371602305

akabiru committed:
Merge branch 'dev' into implementation/50212-convert-the-edit-page-to-primer-design

akabiru committed:
Merge branch 'dev' into implementation/50212-convert-the-edit-page-to-primer-design

akabiru committed:
fix[Op#50742] Replace `StorageErrorPresenter` with new contract x model errors

https://community.openproject.org/work_packages/50742

akabiru committed:
fix[Op#50212]: Reinstate new storage button component template

HDinger commented on modules/storages/app/components/storages/admin/storage_view_component.html.erb:
```diff
@@ -0,0 +1,165 @@
+<%=
+  render(Primer::Beta::BorderBox.new) do |component|
+    component.with_header(color: :default) do
+      grid_layout('op-storage-view--header', tag: :div, align_items: :center) do |grid|
```
Why do you need a grid if there is only one element inside anyways?

akabiru replied:
yeah, you're right-we don't need it 🧹

HDinger reviewed (commented):
Hi @akabiru 

nice job, it looks really awesome! I did not have a complete storage setup, so I could test only half of the things. It might be that some of the issues I found are a result of this half setup. 

- [x] When I try to add a new storage, I get a black screen with a weird message of a missing parameter as soon as I press "Save":
<img width="471" alt="Bildschirmfoto 2023-10-25 um 19 06 05" src="https://github.com/opf/openproject/assets/7457313/63be3e9e-5a5c-4589-b16b-c4bac8821a0b">

- [ ] When trying to refresh an incomplete OAuthApplication via the arrows on the right, I get a routing error.
<img width="200" alt="Bildschirmfoto 2023-10-25 um 19 06 44" src="https://github.com/opf/openproject/assets/7457313/11548b52-0390-43c0-9ac5-56f2bbb38a88">

- [ ] On mobile, there are some overlappings
<img width="200" alt="Bildschirmfoto 2023-10-25 um 19 07 44" src="https://github.com/opf/openproject/assets/7457313/cc5bcb7f-05d3-4c2c-9dfc-dc914c0ea6bc">

- [x] Compared to the visuals, there is no breadcrumb shown for me. I asuume that is part of the other PR, right?

As I said, I don't know if these are actual issues or come due to my hacked setup. If you have problems reproducing the issues, I can also show you tomorrow.

HDinger reviewed (changes requested):
Hi @akabiru 

I was able to test it better now. Some additional remarks I found:

- [x] Deleting the storage results in a routing error,
- [x] When saving the automatically saved folders, the section does not collapse, but stays open

akabiru commented:
> Hi @akabiru
> 
> I was able to test it better now. Some additional remarks I found:
> 
>     * [ ]  Deleting the storage results in a routing error,

Thanks @HDinger, the delete problem is a known issue unfortunately captured separately in (https://community.openproject.org/projects/openproject/work_packages/50739) - but I understand the expecation for it to work in this PR- I'll look into it 🔧 

>     * [ ]  When saving the automatically saved folders, the section does not collapse, but stays open

Will reach out directly as I haven't observed this problem

akabiru committed:
fix[#50739] DELETE storage button fails and is converted to a GET request

https://community.openproject.org/work_packages/50739

akabiru committed:
chore[Op#50212]: remove unneeded header grid layout

akabiru committed:
fix[Op#50212]: check for params presence, pending FormObject implementation

akabiru reviewed (commented):
Thanks for the review @HDinger 🙌🏾 

_Resolved_

- [x] When I try to add a new storage, I get a black screen with a weird message of a missing parameter as soon as I press "Save":
<img width="471" alt="Bildschirmfoto 2023-10-25 um 19 06 05" src="https://github.com/opf/openproject/assets/7457313/63be3e9e-5a5c-4589-b16b-c4bac8821a0b"> https://github.com/opf/openproject/pull/13866/commits/c9d82979d3d3265a49db4502727c7d31cfdb65ea

- [x] When trying to refresh an incomplete OAuthApplication via the arrows on the right, I get a routing error.
<img width="200" alt="Bildschirmfoto 2023-10-25 um 19 06 44" src="https://github.com/opf/openproject/assets/7457313/11548b52-0390-43c0-9ac5-56f2bbb38a88">  _manually created storage that did not have an oauth application_
- [x] Why do you need a grid if there is only one element inside anyways? https://github.com/opf/openproject/pull/13866/commits/11188f2d70585c03bcb6883e38114559cfc94765

🚧 _Separate Work Packages: followup work_

- [x] On mobile, there are some overlappings
<img width="200" alt="Bildschirmfoto 2023-10-25 um 19 07 44" src="https://github.com/opf/openproject/assets/7457313/cc5bcb7f-05d3-4c2c-9dfc-dc914c0ea6bc"> https://community.openproject.org/wp/50762

- [x] Compared to the visuals, there is no breadcrumb shown for me. I asuume that is part of the other PR, right? ✅  https://github.com/opf/openproject/pull/13962

aaron-contreras commented on modules/storages/app/views/storages/admin/storages/edit.html.erb:
```diff
@@ -37,18 +37,24 @@ See COPYRIGHT and LICENSE files for more details.
-          I18n.t('button_delete')
+        ) do |_form|
+          render(
+            Primer::Beta::Button.new(
+              scheme: :danger,
+              size: :medium,
+              type: :submit,
+              aria: { label: I18n.t("storages.label_delete_storage") },
+              data: { confirm: I18n.t('storages.delete_warning.storage') },
+              test_selector: 'storage-delete-button'
+            )
+          ) do |button|
+            button.with_leading_visual_icon(icon: :trash)
+            I18n.t('button_delete')
+          end
```
⭐️

HDinger reviewed (commented):
Hi @akabiru 

The changes look good to me. 👍 I found one other issue, but I am fine with moving that to a separate ticket if you want to:
* When I change the name of a storage, the name is not updated on the page header (only with a hard reload). However, as there is no success message or anything to indicate that the change was accepted, it felt like a bug.

-------------

Additionally, I received another routing error when trying to change the name of one of my storages. However, it did happen only for this specifc storage, so that might be due to my setup.

https://github.com/opf/openproject/assets/7457313/a1aa13f9-c39a-43fb-b9d3-a520a905b3fd

akabiru commented:
> Hi @akabiru
> 
> The changes look good to me. 👍 I found one other issue, but I am fine with moving that to a separate ticket if you want to:
> 
>     * When I change the name of a storage, the name is not updated on the page header (only with a hard reload). However, as there is no success message or anything to indicate that the change was accepted, it felt like a bug.

Thanks @HDinger , the header name update is WIP separately in https://community.openproject.org/wp/50738

> 
> Additionally, I received another routing error when trying to change the name of one of my storages. However, it did happen only for this specifc storage, so that might be due to my setup.

Hmm, I suspect this is due to a manually created storage- as `https://example.com` is not a valid nextcloud URL, and would fail if created via the Form/API I will nonetheless track it as a bug to handle more gracefully. _**Confirmed** if the storage is created manually via console, the dependent oauth application relation is not created which causes the crash above_

![Screenshot 2023-10-27 at 9 51 05 AM](https://github.com/opf/openproject/assets/17295175/a0627af9-01d9-4be2-b435-ad08825814e8)

HDinger reviewed (approved):
👍 🌟
